packages/sdk/src/cross-chain-provider.ts

```diff
@@ -422,7 +422,21 @@ export class CrossChainProvider implements ICrossChainProvider {
   public async estimateL2MessageGasLimit(
     message: MessageLike
   ): Promise<BigNumber> {
-    throw new Error('Not implemented')
+    const resolved = await this.toCrossChainMessage(message)
+
+    // L2 message gas estimation is only used for L1 => L2 messages.
+    if (resolved.direction === MessageDirection.L2_TO_L1) {
+      throw new Error(`cannot estimate gas limit for L2 => L1 message`)
+    }
+
+    const estimate = await this.l2Provider.estimateGas({
+      from: resolved.sender,
+      to: resolved.target,
+      data: resolved.message,
+    })
+
+    // Return the estimate plus a buffer of 20% just in case.
+    return estimate.mul(120).div(100)
```

Comment:
Were you noticing issues without the buffer? 20% seems pretty large as a buffer

Reply:
What do you think about an optional argument `buffer` that defaults to 20 or some other value

Reply:
Hmmm could do. I was just worried about people using this for estimations and then having something be unaccounted for where the buffer would've been necessary. I'll do an optional `buffer` argument.

Reply:
Added a buffer variable